src/Fix.js

```diff
@@ -0,0 +1,120 @@
+ * to remediate the issue.
+ * 
+ * A subclass of Fix can contain any properties - it is up to
+ * the corresponding Fixer (i.e. with a matching {@link Fix.type})
+ * to interpret them appropriately and apply them
+ * to the IntermediateRepresentation.
+ * 
+ * Due to that, a recommended approach is that the Fixer
+ * should also be a Fix factory, so that the Rule
+ * could only express the Fix instructions with
+ * capabilities that the Fixer provides.
+ * 
+ * Example scenario:
+ * 
+ * Take an IntermediateRepresentation with type `string`
+ * which stores content of a file verbatim:
+ * ```text
+ * birds are not real birds are not real birds are not real
+ * ```
+ * and a Rule (type `string`) which ensures that the word `birds`
+ * should always be surrounded by quotes.
+ * 
+ * Rule would produce the following results:
+ * 1. birds are not quoted (span 0 - 5)
+ * 2. birds are not quoted (span 19 - 24)
+ * 3. birds are not quoted (span 38 - 43)
+ * 
+ * To fix each of them, quotes need to be inserted:
+ * 1. fix: insert quote at 0 and 5
+ * 2. fix: insert quote at 19 and 24
+ * 3. fix: insert quote at 38 and 43
+ * 
+ * If there are no Fix and Fixer with type `string`,
+ * they could be implemented like in the following example:
+ * ```ts
+ * class StringFix extends Fix {
+ *     commands: StringFixCommand[];
+ * }
+ * class StringFixer {
+ *     applyFixes(representation: IntermediateRepresentation<string>, fixes: StringFix[]): void
+ *     createFix(commands: StringFixCommand[]): StringFix
+ *     insertStringAt(position: number, content: string): InsertCommand
+ *     removeStringAt(position: number, length: number): RemoveCommand
+ * }
+ * type StringFixCommand = InsertCommand | RemoveCommand;
+ * type InsertCommand = {
+ *     command: 'INSERT',
+ *     position: number,
+ *     content: string
+ * }
+ * type RemoveCommand = {
+ *     command: 'REMOVE',
+ *     position: number,
+ *     length: number
+ * }
+ * ```
+ * 
+ * Rule should then accept an instance of a StringFixer,
+ * and for each produced result provide a fix:
+ * ```
+ * new Result({
+ *     // ...
+ *     fix: fixer.createFix([fixer.insertStringAt(0, "\""), fixer.insertStringAt(5, "\"")])
+ * });
+ * ```
+ * 
+ * So that the linter could call
+ * ```
+ * fixer.applyFixes(ir, results.map(r => r.fix))
+ * ```
+ */
+export class Fix {
+    /**
+     * Unique identifier which allows to dynamically match 
+     * the Fix to its corresponding Fixer and IntermediateRepresentation
+     * onto which it should be applied.
+     * 
+     * Subclass must define this property.
+     * @readonly
+     * @abstract
+     * @type {string}
+     */
+    // @ts-expect-error: subclass must define this property
+    type;
+
+    /** 
+     * If the fix had been applied by the fixer.
+     * Fixer is expected to set this flag after it had applied the fix.
+     * 
+     * @type {boolean}
+     */
+    applied = false;
+}
+
```

Comment:
I was thinking about it but decided against it in the end, because the Fix class is already an abstract container for any data that a given (matching) fixer might need to perform a fix - so there is no point in introducing an additional abstract container class FixCommand which has literally no properties of its own.

You define additional properties of a Fix when declaring a `class SpecificFix extends Fix`, so then within the fixer.apply You will know about them anyway, since You expect that the fixer.fix receives only instances of SpecificFix.